src/Microsoft.ServiceHub.Framework/AsyncNamedPipeClientStream.cs

```diff
@@ -0,0 +1,174 @@
+	{
+		private const int SECURITYSQOSPRESENT = 0x00100000;
+		private const int GENERICREAD = unchecked((int)0x80000000);
+		private const int GENERICWRITE = 0x40000000;
+		private readonly string pipePath;
+		private readonly TokenImpersonationLevel impersonationLevel;
+		private readonly PipeOptions pipeOptions;
+		private readonly PipeDirection direction;
+		private readonly int access;
+
+		/// <summary>
+		/// Initializes a new instance of the <see cref="AsyncNamedPipeClientStream"/> class.
+		/// </summary>
+		/// <param name="serverName">Pipe server name.</param>
+		/// <param name="pipeName">Pipe name.</param>
+		/// <param name="direction">Communication direction.</param>
+		/// <param name="options">Pipe options.</param>
+		/// <param name="impersonationLevel">Impersonation level.</param>
+		internal AsyncNamedPipeClientStream(
+			string serverName,
+			string pipeName,
+			PipeDirection direction,
+			PipeOptions options,
+			TokenImpersonationLevel impersonationLevel = TokenImpersonationLevel.None)
+			: base(direction, 4096)
+		{
+			Requires.NotNullOrEmpty(serverName, nameof(serverName));
+			Requires.NotNullOrEmpty(pipeName, nameof(pipeName));
+
+			this.pipePath = $@"\\{serverName}\pipe\{pipeName}";
+			this.direction = direction;
+			this.pipeOptions = options;
+			this.impersonationLevel = impersonationLevel;
+			this.access = this.GetAccess();
+		}
+
+		/// <summary>
+		/// Connects pipe client to server.
+		/// </summary>
+		/// <param name="cancellationToken">Cancellation token.</param>
+		/// <param name="maxRetries">Maximum number of retries.</param>
+		/// <param name="retryDelayMs">Milliseconds delay between retries.</param>
+		/// <returns>A task representing the establishment of the client connection.</returns>
+		/// <exception cref="TimeoutException">Thrown if no connection can be established.</exception>
+		public async Task ConnectAsync(
+			CancellationToken cancellationToken,
+			int maxRetries,
+			int retryDelayMs)
+		{
+			var errorCodeMap = new Dictionary<int, int>();
+			int retries = 0;
+			while (!cancellationToken.IsCancellationRequested)
+			{
+				if (retries > maxRetries || this.TryConnect(ref errorCodeMap))
+				{
+					break;
+				}
+
+				retries++;
+				await Task.Delay(retryDelayMs, cancellationToken).ConfigureAwait(false);
+			}
+
+			if (retries > maxRetries || !this.IsConnected)
+			{
+				throw new TimeoutException($"Failed with errors: {string.Join(", ", errorCodeMap.Select(x => $"(code: {x.Key}, count: {x.Value})"))}");
+			}
+		}
+
+		[DllImport("kernel32.dll", EntryPoint = "CreateFileW", CharSet = CharSet.Unicode, SetLastError = true)]
+		private static extern SafePipeHandle CreateNamedPipeClient(
+			string lpFileName,
+			uint dwDesiredAccess,
+			uint dwShareMode,
+			IntPtr secAttrs,
+			uint dwCreationDisposition,
+			uint dwFlagsAndAttributes,
+			IntPtr hTemplateFile);
+
+		private bool TryConnect(ref Dictionary<int, int> errorCodeMap)
```

Comment:
🤔 Why are you using `ref` here? `Dictionary<K,V>` is a mutable reference type, and you aren't reassigning the variable, so it just makes it slightly harder to call the method to use this modifier here.